keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,1128 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes introducing a new field, `routingPreference`, to the
+Kubernetes Service spec. It will supersede the functionality currently provided
+by the `service.kubernetes.io/topology-mode` annotation and it’s precursor
+`topologyKeys` field (which has been deprecated since Kubernetes 1.21)
+
+## Motivation
+
+To be able to understand the motivations for introducing this new field, it’s
+important to understand the precursors that shaped this KEP.
+
+
+### Topology Keys
+
+* This early mechanism allowed users to specify ordered routing preferences (e.g.,
+prioritize local nodes, then zone, then anywhere). 
+
+* **Limitations:** However, it lacked flexibility for implementations to make smart
+  decisions on its own, like incorporating feedback-based routing optimizations
+  (e.g., avoiding overloaded endpoints). The name `topologyKeys` might be
+  perceived as overly restrictive, given the desire to incorporate non-topology
+  factors into routing decisions.
+
+### Topology Aware Routing and InternalTrafficPolicy
+
+TopologyAwareRouting together with InternalTrafficPolicy were meant to be the
+successors of `topologyKeys` and allow implementations to be more flexible.
+
+* TopologyAwareRouting: 
+  * Exposes the annotation service.kubernetes.io/topology-mode. When this
+    annotation is set to Auto, an implementation specific heuristic is used to
+    route the traffic. 
+  * **Goal:** The aim with Auto was to allow implementations to be as smart as
+    possible and choose what it considered to be the best routing criteria. The
+    field didn’t support any other value by design with the thought that the
+    implementation should be able to make the best possible choice for the user.
+  * **Limitation:** While designed for intelligent routing, this mode offers less
+    user control and can be less predictable. Some users prioritize
+    predictability for performance optimization, even when accepting potential
+    endpoint overload in certain situations.
+
+* InternalTrafficPolicy 
+  * The spec.internalTrafficPolicy field with the "Local" setting restricts
+    traffic to endpoints on the same node as the originating Pod. 
+  * **Goal:** With the deprecation of topologyKeys, one of the driving reasons was
+    to satisfy the semantic use cases of strictly routing the traffic to the
+    same node.
+  * **Limitation:** Lacks failover; traffic is dropped if no local endpoint exists.
+
+Note that while the initial proposal of InternalTrafficPolicy proposed a
+PreferLocal policy, it was dropped later on. This meant that now
+TopologyAwareRouting in conjunction with InternalTrafficPolicy didn’t exactly
+allow users to express a much desired use case from topologyKeys which is
+"prefer node-local, failover to same zone, then route anywhere"
```

Comment:
Although this is all 100% correct, I don't think this KEP really addresses that?